common/latch_test.cpp

```diff
@@ -0,0 +1,111 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "common/latch.h"
+
+#include <gtest/gtest.h>
+
+#include <chrono>
+#include <thread>
+#include <vector>
+
+namespace Carbon {
+namespace {
+
+// Basic test for the Latch.
+TEST(LatchTest, Basic) {
+  Latch latch;
+  Latch::Handle handle = latch.Init();
+  // Dropping a copy doesn't satisfy the latch.
+  Latch::Handle handle_copy = handle;
+  EXPECT_FALSE(std::move(handle_copy).Drop());
+  // Can create more copies even after we start dropping.
+  Latch::Handle handle_copy2 = handle;
+  EXPECT_FALSE(std::move(handle_copy2).Drop());
+  // Now drop the last handle.
+  EXPECT_TRUE(std::move(handle).Drop());
+}
+
+// Tests that the on-zero callback is called.
+TEST(LatchTest, OnZeroCallback) {
+  Latch latch;
+  bool called = false;
+  Latch::Handle handle = latch.Init([&] { called = true; });
+  Latch::Handle handle2 = handle;
+  Latch::Handle handle3 = handle;
+
+  EXPECT_FALSE(called);
+  ASSERT_FALSE(std::move(handle).Drop());
+  EXPECT_FALSE(called);
+  ASSERT_FALSE(std::move(handle2).Drop());
+  EXPECT_FALSE(called);
+  ASSERT_TRUE(std::move(handle3).Drop());
+  EXPECT_TRUE(called);
```

Comment:
```suggestion
  EXPECT_FALSE(std::move(handle).Drop());
  EXPECT_FALSE(called);
  EXPECT_FALSE(std::move(handle2).Drop());
  EXPECT_FALSE(called);
  EXPECT_FALSE(std::move(handle3).Drop());
  EXPECT_TRUE(called);
```
The general practice AIUI is to use ASSERT conditions to protect against crashes but otherwise allow the test to proceed as later failures can provide additional useful information.